extensions/script-libraries/extension-utils/extension-utils.livecodescript

```diff
@@ -0,0 +1,1634 @@
+         end if
+         break
+      case "Win32"
+         put "dll" into tFileExtension
+         break
+      case "Linux"
+         put "so" into tFileExtension
+         break
+   end switch
+   
+   local tExternalFilename
+   if the environment is "server" then
+      put "server-" after tExternalFilename
+   end if
+   put pExternal after tExternalFilename
+   put "." & tFileExtension after tExternalFilename
+   
+   if there is a file (tHomeFolder & slash & "externals" & slash & \
+         tExternalFilename) then
+      put true into sExternalsA[pExternal]
+      exit __EnsureExternal
+   end if
+   
+   put specialFolderPath("engine") & slash before tExternalFilename
+   set the externals of the templateStack to tExternalFilename
+   
+   create stack pExternal && "External"
+   start using it
+   if the externalCommands of it is empty then
+      throw "cannot load external" && pExternal
+   else
+      put true into sExternalsA[pExternal]
+   end if
+end __EnsureExternal 
+
+private on __addToDependencies pDependee, pDependent, @xDependencies
+   # AL-2015-04-13: [[ Bug 15216 ]] Use keys of array to remove duplicates
+   put true into xDependencies[pDependee][pDependent]
+end __addToDependencies
+
+private on __addToList pNode, @xList
+   if xList is empty then
+      put pNode into xList
+   else
+      put return & pNode after xList
+   end if
+end __addToList
+
+private on __removeFromList pNode, @xList
+   get lineOffset(pNode, xList)
+   if it is not 0 then
+      delete line it of xList
+   end if
+end __removeFromList
+
+private on __visitNode pDependencies, pNode, @xUnmarked, @xMarked, @xTemporaryMarked, @xOrder
+   if pNode is among the lines of xTemporaryMarked then
+      return "Error"
+   end if
+   if pNode is among the lines of xMarked then
+      # AL-2015-04-13: [[ Bug 15216 ]] Remove from unmarked if this was already marked
+      __removeFromList pNode, xUnmarked
+      return ""
+   end if
+   __addToList pNode, xTemporaryMarked
+   repeat for each key tNode in pDependencies[pNode]
+      __visitNode pDependencies, tNode, xUnmarked, xMarked, xTemporaryMarked, xOrder
+      if the result is not empty then
+         return "Error"
+      end if
+   end repeat
+   __addToList pNode, xMarked
+   __removeFromList pNode, xUnmarked
+   __removeFromList pNode, xTemporaryMarked
+   __addToList pNode, xOrder
+end __visitNode
+
+private function __dependencyOrder pDependencies, pList
+   local tUnmarked, tTemporaryMarked, tMarked, tOrder, tToVisit
+   put pList into tUnmarked
+   repeat while tUnmarked is not empty
+      put any line of tUnmarked into tToVisit
+      __visitNode pDependencies, tToVisit, tUnmarked, tMarked, tTemporaryMarked, tOrder
+      if the result is not empty then
+         throw "Circularity in dependencies starting from" && tToVisit
+         return empty
+      end if
+   end repeat
+   return tOrder
+end __dependencyOrder
+
+private function __extensionIsBuiltin pID
+   switch pID
```

Comment:
@livecodeali this seems like something we might want to add an enhancement request for a function